### DISCUSSION TITLE
how can I get the embeddings associated with each prediction (or box)?

### DISCUSSION THREAD

**Discussion by pax7** [ultralytics/ultralytics#22625] [Q&A] [answered]
how can I get the embeddings associated with each prediction (or box)?

so basically I want do do

```python

model.predict()

for box in boxes:
  emb = box.embeddings

```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pax7, thanks for opening a discussion and for your interest in Ultralytics 🚀! This is an automated response—an Ultralytics engineer will follow up shortly to assist you.

For getting embeddings associated with each prediction/box, please share a brief code snippet showing how you’re running inference (Python or CLI), the exact model you’re using (e.g., YOLO11 variants), and a sample output you currently see. This will help us guide you on the best approach for your use case.

We also recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report related to embeddings not appearing or mismatched with boxes, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (e.g., training a model to output embeddings), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Craniace** [comment]
```
from ultralytics import YOLO
import torch

model = YOLO("yolov8n.pt")

# Forward pass and get intermediate features
image = "image.jpg"
results = model(image)

# Access model backbone outputs
features = model.model.model[0](torch.randn(1, 3, 640, 640))
print([f.shape for f in features])  # feature maps
```

You can use yolo for it.

**Reply by pax7** [reply]
thanks for responding - the code above is throwing an error - but also, I am not seeing how I can get the embeddings arrays that correspond to each detection

**Reply by glenn-jocher** [reply, marked as answer]
@pax7 The public `Results` object doesn’t attach per-box embeddings; the supported way is to crop each detected box and run those crops through YOLO as an encoder via `model.embed(...)`:

```python
from ultralytics import YOLO
from ultralytics.utils.plotting import save_one_box

detector = YOLO("yolo11n.pt")
res = detector.predict("image.jpg", verbose=False)[0]

# crop each detection
crops = [save_one_box(xyxy, res.orig_img, save=False) for xyxy in res.boxes.xyxy.cpu()]

# get one embedding per crop (defaults to the penultimate layer; you can override with embed=[layer_idx])
encoder = YOLO("yolo11n.pt")
embs = encoder.embed(crops)  # list[Tensor], aligns 1:1 with res.boxes
```

API details are in the [YOLO.embed reference](https://docs.ultralytics.com/reference/engine/model/#embed) and, for advanced/internal use, the detector has a helper to derive features per detection from feature maps ([get_obj_feats](https://docs.ultralytics.com/reference/models/yolo/detect/predict/#get_obj_feats)); if you go that route you’ll be working below the public API.

**Reply by pax7** [reply]
thanks, this solved it for me!

**Reply by glenn-jocher** [reply]
Great—thanks for confirming! Please mark the earlier reply as the answer for visibility, and open a new discussion if anything else comes up.
